### ISSUE TITLE
Optimize number of database queries triggered by nested Prisma Client requests

### ISSUE THREAD

**Issue by nikolasburk** [prisma/prisma#4997]
Currently Prisma Client (or more precisely: the underlying query engine) generates multiple database queries for a nested read API call. Consider this example:

Assume there is a nested read API call that goes 3 levels deep, e.g.: fetch user with posts with comments.

This is what's happening under the hood:

1. Prisma Client generates a single nested GraphQL query and sends it to query engine
2. The query engine receives the query
3. The query engine fetches the user
4. The query engine fetches the user's posts
5. The query engine fetches the posts' comments
6. The query engine packages up the GraphQL response and sends it back to Prisma Client

If there aren't any filter arguments provided for the posts or the comments, this will result in 3 database queries (possibly more if there are filter arguments).

It is not yet clear if/how this will be optimized in the future. This issue is used for discussion and tracking the progress of the potential optimization.

**Comment by arcticmatt** [comment]
> The total cost of executing a complex join is often higher than executing multiple simpler queries.

Are you accounting for network latency?

**Comment by dpr-dev** [comment]
small example

Prisma -- 3 queries for insert & select -- 1.52s execution time 
EF Core -- same queries with the same db -- 200ms execution time 


It's look like the prisma should have an option to query data with joins

**Comment by chrbala** [comment]
> > The total cost of executing a complex join is often higher than executing multiple simpler queries.
> 
> Are you accounting for network latency?

I evaluated this here: #12582

**Comment by lightningspirit** [comment]
> > The total cost of executing a complex join is often higher than executing multiple simpler queries.
> 
> Are you accounting for network latency?

It's not only about network latency. Current enterprise grade DRBM engines are optimised for matrix operations, having built-in joins as one example. 

For relatively small databases (hundreds to few thousands of rows), and depending on the query expression, the penalty of having multiple simpler queries vs one single complex expression is most of the time negligible.

However, looking at bigger sets, you'll start to see a higher workload on the engines as well as high memory usage and CPU peaks and dependent applications will start to perform much worse. Taking MySQL for instance, this is notorious when performing tests on tables with millions of records. Not only to mention the fact that internal engine's caching algo could play an important role for most of the query expressions on CRUD applications.

Don't want to start a flame war 🔥 here, but I was using TypeORM and migrated to Prisma because of its poor query optimization and bugs found (and actually reported). I actually love Prisma's direction and core stability but, IMO I think looking at nested query otimization is very important for Prisma to take a leading step on ORM world.

**Comment by abenhamdine** [comment]
> [on 31 Jul 2019](https://github.com/prisma/prisma/issues/4997#issuecomment-516806095)

> The total cost of executing a complex join is often higher than executing multiple simpler queries. This is why the Prisma query engine currently defaults to multiple simple queries in order to optimise overall throughput of the system.

This is not true, according to my experience, at least with sophisticated RDBM like postgres (and probably MSQL Server, Oracle, even MySql I think)
unless you are only talking of the part of the execution in Prisma, and not the overall query execution ?

As a CTO of a company with a product run millions of complexe queries every day, I've always found that fewer queries are almost always much faster.

Each query adds an overhead, at db level and driver level : 
- work of parsing and planning multiple queries is already more than any possible gain in most cases.
- on client , parsing/combining the results
- latency, but not only :  multiple round trips to the database server, means multiple times the network latency and transaction overhead
- more data transfered between db server and client
- with a single query, the db server can make optimizations, split the plan to perform parallelism on some operations, etc. You can't reproduce these kind of complexe optimizations on client side : for example, the driver node-postgres does not support pipelining, so it's prevent you to really send parallel queries
- etc..
- 
I observed many times that the network latency alone can kill the performance.

**Comment by AndrewSouthpaw** [comment]
Agreed that this design ignores considerations of network latency. What we've found experimentally is that doing multiple simple queries in series turns a 100ms query into a 400ms one.

**Comment by arcticmatt** [comment]
Curious if there are any updates on this?

When you include a lot of stuff, the number of queries Prisma sends can get quite large

I'm not totally sure if a JOIN would be faster, but it would be nice to have at least have control so we can see if it would be faster for some of our queries

<img width="1481" alt="image" src="https://user-images.githubusercontent.com/4356178/218235317-543faaf1-9e06-4935-b59e-6ffebae5cc51.png">

**Comment by zadamg** [comment]
> Curious if there are any updates on this?
> 
> When you include a lot of stuff, the number of queries Prisma sends can get quite large
> 
> I'm not totally sure if a JOIN would be faster, but it would be nice to have at least have control so we can see if it would be faster for some of our queries
> 
> <img alt="image" width="1481" src="https://user-images.githubusercontent.com/4356178/218235317-543faaf1-9e06-4935-b59e-6ffebae5cc51.png">

I know the great team at prisma has probably already rushed this critical issue to the top of the backlog, but just wanted to add my voice to the pile:

1 query = 1 query.  If this cannot be corrected, then this ORM isn't very good for anything relational. It's basically a postgres wrapped NoSQL because anything else will cost, 2x, 3x, 4x, 5x, while also incurring latency.  

Back to the drawing board, folks!

**Comment by brecht-derooms** [comment]
I'm personally in favor for one query until it doesn't work anymore and have personally encountered a few cases where one query failed in ORMs like EF core but these were exceptions. Like what @abenhamdine testified, from what I've seen in production a single query will be almost always faster because you are using the relational database for what it's meant to be used as @lightningspirit said.   

I did have to debug the occasional slow query in tools like EF Core, why andhow slow it becomes depends on:
- **the amount of columns**: given that Prisma is GraphQL inspired with the 'don't overfetch philosphy' it can potentially  do better than something like EF core where all fields are retrieved, whether you need them or not. Since, in theory, youcan select which fields you need this reduces the problem greatly, how that works out with typed results adds probably some complexity. I had cases that went wrong in EF core resulting in 300 columns which would easily be fixed by only fetching the fields we care about. Maybe every table has created_at/updated_at/updated_by but you don't need any of them in this specific query for example (we had quite a lot of those bookkeeping fields). 
- **the cardinality of your joins**: this will determine how many times a column's field is duplicated due to one-to-many or many-to-many joins as explained by guerrerocarlos. If your joins are one-to-one then a join will always be faster than query splitting. In some databases like Postgres, you could aggregate the rows with array_agg which removes the duplication, I haven't been able to test whether this is more performant. 
- **ordering and other ORM requirements**: some ORMS then order on the full set of columns for pagination reasons or others which typically results in a big chunk of extra time. 
- **silly joins due to not enough knowledge**: sometimes there are links which could be simplified but the ORM has not enough knowledge to do so resulting in too much duplicates. 

Whether one or the other is faster depends on so many factors/variables.  So an algorithm to decide when to split queries would have to take into account how many times the amount of rows will be multipled for each join, how many columns that were already there (that will be duplicated) for each join, taking into account filters etc, taking into account the added network latency that query splitting  would bring when doing multiple queries that have to wait for each other, somethingt that is barely noticeable locally but imho can't be underestimated in real environments (something that my former employers at Fauna were very much invested into). Given this I would prefer not to try to be smart and give users an option that we can pass to enable query splitting, ideally being able to say: "split at this specific join". 

I also have very high hopes for Prisma since after using an ORM or 6-7 it's the best experience but this particular part is my least favorite part. Prisma has the potential to rule them all IMO.

**Comment by janpio** [comment]
@brecht-derooms We fully agree. We are working on query performance and this will certainly be high on our list to consider.

**Comment by Weakky** [comment]
Hey folks,

We’re excited to let you know that we’ve just added support for `JOIN`s to resolve relations in [Prisma 5.7.0](https://github.com/prisma/prisma/releases/tag/5.7.0) for PostgreSQL and CockroachDB, under the `relationJoins` preview feature.

Once you enable that feature, `JOIN`s will be used to resolve one-to-one, one-to-many and many-to-many relations. Most of the queries listed on [this documentation page](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-reads) will essentially now use joins. This means that, when possible, we now resolve your query in a single round-trip as opposed to the N round-trips we used to make for each relation.

I say *when possible*, because we fallback to the old query-based mechanism when you:

- Use `distinct`
- Use `cursor`
- Use `_count` in your selection set

This is the **very first step** to bring `JOIN`s to Prisma for querying relations. We plan on removing the current limitations and adding support for other databases in future releases.

Changing the way Prisma resolves relations is not a trivial change and is likely not going to work 100% on day one, so please bear with us as we continuously improve the feature, both in terms of bugs and performance improvements on the SQL queries we generate.

If you are curious, we have provided a bunch of Prisma query examples together with their associated SQL query output in our [latest release notes](https://github.com/prisma/prisma/releases/tag/5.7.0)

Finally, we have a dedicated [preview feature feedback issue here](https://github.com/prisma/prisma/discussions/22288) if you have any thoughts about it.

Thank you all for your patience!

**Comment by aqrln** [comment]
Hey everyone! As a follow up to the previous comment by @Weakky, in Prisma 5.10.0 we have released support for MySQL 8.0.14 and newer in the `relationJoins` preview feature. MySQL currently still falls back to the old strategy if a query contains `distinct` or `cursor`. Please try it out and let us know your feedback 🙏

**Comment by igalklebanov** [comment]
Hey 👋 

Closing this as it's a thing of the past with latest major version/s.
